server/controllers/classifier.js

```diff
@@ -0,0 +1,179 @@
+                token: tokenMatched.token,
+                type: type,
+                matchedValue: tokenMatched.type,
+                distance: tokenMatched.distance
+            };
+        } else if (lookahead && this.state.currentToken <= this.state.tokenHolders.length - 2) {
+            let nextTokenHolder = this.state.tokenHolders[this.state.currentToken + 1];
+            tokenMatched = this.getMostLikelyMatch(tokenHolder.token + " " + nextTokenHolder.token, valueRange);
+            if (tokenMatched && tokenMatched.distance < tokenHolder.distance) {
+                performedLookahead = true;
+                //replace the currentToken
+                this.state.tokenHolders[this.state.currentToken] = {
+                    token: tokenMatched.token,
+                    type: type,
+                    matchedValue: tokenMatched.type,
+                    distance: tokenMatched.distance
+                };
+            }
+        }
+        this.nextAction(performedLookahead);
+    }
+
+    /**
+     * gets the most likely match for a token out of an array of values
+     * @param token token to match
+     * @param possibleTypes possible values which it could match
+     * @returns {*}
+     */
+    getMostLikelyMatch(token, possibleTypes) {
+        let ratedTypeAffiliation = [];
+        possibleTypes.forEach(type => {
+            let distance;
+            if (type.synonyms && type.synonyms.length != 0) {
+                //get best fitting synonym
+                let labelVariation = [];
+                labelVariation [0] = type.label;
+                labelVariation = labelVariation.concat(type.synonyms);
+                let bestSynonym = this.getMostLikelyMatch(token, this.createLabelSynonymStructure(labelVariation));
+                distance = bestSynonym.distance;
+            } else {
+                distance = this.getLevenshteinDistance(token, type.label);
+            }
+            ratedTypeAffiliation.push({
+                "type": type.label,
+                "distance": distance,
+                "token": token
+            });
+        })
+
+        let minDistance = Math.min.apply(Math, ratedTypeAffiliation.map(ratedType => ratedType.distance));
+
+        let mostLikelyMatch = ratedTypeAffiliation.find((ratedType) => ratedType.distance == minDistance);
+        //console.log("Token: " + token + " Distance: " + mostLikelyMatch.distance + " matched to " + mostLikelyMatch.type)
+        return mostLikelyMatch;
+    }
+
+
+    getLevenshteinDistance(token, label) {
+        let weight = 5 / (token.length / 2);
+        let distance = natural.LevenshteinDistance(token, label, {
+            insertion_cost: 2 * weight,
+            deletion_cost: 2 * weight,
+            substitution_cost: weight
+        });
+        return distance;
+    }
+}
+
+Classifier.staticWords = {
+    column: "Column",
+    operation: "Operation",
+    chartType: "ChartType"
+};
+
+Classifier.maxDistance = 5;
```

Comment:
With a distance of 5, it detects "race" as "make" and does not go to the match column part. It would be best if we search through all search functions and look if there is a more exact match. But that would need a lot of restructuring I think..

Reply:
yes, but adding and removing letters from a word costs already 2* weight which is 5/(token.length/2) so I found it okay,  e.g. "and" is not mapped to "age"